pkg/engine/task/task_apply.go

```diff
@@ -0,0 +1,137 @@
+package task
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1alpha1"
+	"github.com/kudobuilder/kudo/pkg/util/health"
+	apiextv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	apijson "k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ApplyTask will apply a set of given resources to the cluster. See Run method for more details.
+type ApplyTask struct {
+	Name      string
+	Resources []string
+}
+
+// Run method for the ApplyTask. Given the task context, it renders the templates using context parameters
+// creates runtime objects and kustomizes them, and applies them using the controller client. Finally,
+// resources are checked for health.
+func (at ApplyTask) Run(ctx Context) (bool, error) {
+	// 1. - Render task templates -
+	rendered, err := render(at.Resources, ctx.Templates, ctx.Parameters, ctx.Meta)
+	if err != nil {
+		return false, fmt.Errorf("%wfailed to render task resources: %v", ErrFatalExecution, err)
+	}
+
+	// 2. - Kustomize them with metadata -
+	kustomized, err := kustomize(rendered, ctx.Meta, ctx.Enhancer)
+	if err != nil {
+		return false, fmt.Errorf("%wfailed to kustomize task resources: %v", ErrFatalExecution, err)
+	}
+
+	// 3. - Apply them using the client -
+	applied, err := apply(kustomized, ctx.Client)
+	if err != nil {
+		return false, err
+	}
+
+	// 4. - Check health for all resources -
+	err = isHealthy(applied, ctx.Client)
+	if err != nil {
+		// so far we do not distinguish between unhealthy resources and other errors that might occur during a health check
+		// an error during a health check is not treated task execution error
+		log.Printf("TaskExecution: %v", err)
+		return false, nil
+	}
+	return true, nil
+}
+
+// apply method takes a slice of k8s object and applies them using passed client. If an object
+// doesn't exist it will be created. An already existing object will be patched.
+func apply(ro []runtime.Object, c client.Client) ([]runtime.Object, error) {
+	applied := make([]runtime.Object, len(ro))
+
+	for _, r := range ro {
```

Comment:
@zen-dog wonder if we could get rid of a bunch of code using https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate ?

Reply:
well our patch logic is kind of special here, I was actually afraid to touch this in my refactoring :D I would PROBABLY try to tackle that in separate PR? But of course up to Aleksey :)

Reply:
:+1: on separate PR for this.

Reply:
+1 to a separate PR.